Some images on www.dandelionchocolate.com fail to load.

mattwoodrow opened the pull request:
#### 5dda919ecebde78601eeac9319eec101f6377124
<pre>
Some images on www.dandelionchocolate.com fail to load.
<a href="https://bugs.webkit.org/show_bug.cgi?id=284860">https://bugs.webkit.org/show_bug.cgi?id=284860</a>
&lt;<a href="https://rdar.apple.com/141520514">rdar://141520514</a>&gt;

Reviewed by Simon Fraser.

This is a bad interaction between the 284780@main changes, and the old
‘simplified repaint’ optimisation.

A style change that results in a compositing layer being created/destroyed
previously wrote new ‘repaint rects’ onto the RenderLayer. There was a comment
complaining about this, as computing repaint rects while layout was dirty was
invalid.

The new code in 284780@main just sets the dirty bit on the layer, so that we
know to recompute repaint rects after the next layout.

If that next layout ends up being a ‘simplified’ one, then the simplified
repaint optimisations applied to the layer positions update and stop the repaint
rects from being computed.

The simplified repaint optimisation isn’t valid when dirty bits are set outside
of the layout immediately preceding the layer position update.

Fix this by removing the simplified repaint optimization, since it should be
superceeded by the layer dirty flags.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::didLayout):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::performLayout):
(WebCore::LocalFrameViewLayoutContext::didLayout):
(WebCore::LocalFrameViewLayoutContext::flushUpdateLayerPositions):
* Source/WebCore/page/LocalFrameViewLayoutContext.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::didVisitSinceLayout const): Deleted.
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::setLayoutIdentifier): Deleted.
(WebCore::RenderElement::layoutIdentifier const): Deleted.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateLayerPositionsAfterStyleChange):
(WebCore::RenderLayer::updateLayerPositionsAfterLayout):
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::clearNeedsLayout):
(WebCore::RenderObject::markContainingBlocksForLayout):
(WebCore::RenderObject::outputRenderObject const):
(WebCore::setIsSimplifiedLayoutRootForLayerIfApplicable): Deleted.

Canonical link: <a href="https://commits.webkit.org/288074@main">https://commits.webkit.org/288074@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ba6e054af5fc893135d90382db730a0792a1c86

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81735 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1261 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35689 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86280 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32731 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83841 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1294 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9081 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63767 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21493 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84805 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/937 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74384 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44053 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/837 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28559 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31184 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29164 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87718 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8975 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6376 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/scripting-1/the-template-element/template-element/template-descendant-body.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72107 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9160 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70204 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71336 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17786 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15421 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14340 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8926 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14458 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8767 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12290 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10575 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr reviewed (approved):
r=me but we should try hard to write a test

mattwoodrow committed:
Some images on www.dandelionchocolate.com fail to load.
https://bugs.webkit.org/show_bug.cgi?id=284860
<rdar://141520514>

Reviewed by Simon Fraser.

This is a bad interaction between the 284780@main changes, and the old
‘simplified repaint’ optimisation.

A style change that results in a compositing layer being created/destroyed
previously wrote new ‘repaint rects’ onto the RenderLayer. There was a comment
complaining about this, as computing repaint rects while layout was dirty was
invalid.

The new code in 284780@main just sets the dirty bit on the layer, so that we
know to recompute repaint rects after the next layout.

If that next layout ends up being a ‘simplified’ one, then the simplified
repaint optimisations applied to the layer positions update and stop the repaint
rects from being computed.

The simplified repaint optimisation isn’t valid when dirty bits are set outside
of the layout immediately preceding the layer position update.

Fix this by removing the simplified repaint optimization, since it should be
superceeded by the layer dirty flags.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::didLayout):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::performLayout):
(WebCore::LocalFrameViewLayoutContext::didLayout):
(WebCore::LocalFrameViewLayoutContext::flushUpdateLayerPositions):
* Source/WebCore/page/LocalFrameViewLayoutContext.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::didVisitSinceLayout const): Deleted.
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::setLayoutIdentifier): Deleted.
(WebCore::RenderElement::layoutIdentifier const): Deleted.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateLayerPositionsAfterStyleChange):
(WebCore::RenderLayer::updateLayerPositionsAfterLayout):
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::clearNeedsLayout):
(WebCore::RenderObject::markContainingBlocksForLayout):
(WebCore::RenderObject::outputRenderObject const):
(WebCore::setIsSimplifiedLayoutRootForLayerIfApplicable): Deleted.

Canonical link: https://commits.webkit.org/288074@main
